File: OOP2Struc3.cpp
#include "pt4.h"
using namespace std;

class TextView
{
	// §¯§¦ §ª§©§®§¦§¯§Á§«§´§¦ §²§¦§¡§­§ª§©§¡§¸§ª§À §¥§¡§¯§¯§°§¤§° §¬§­§¡§³§³§¡
    int x = 0, y = 0;
    int width = 1, height = 1;
public:
    void GetOrigin(int& x, int& y);
    void SetOrigin(int x, int y);
    void GetSize(int& width, int& height);
    void SetSize(int width, int height);
};

void TextView::GetOrigin(int& x, int& y)
{
    x = this->x;
    y = this->y;
}
void TextView::SetOrigin(int x, int y)
{
    this->x = x;
    this->y = y;
}
void TextView::GetSize(int& width, int& height)
{
    width = this->width;
    height = this->height;
}
void TextView::SetSize(int width, int height)
{
    this->width = width;
    this->height = height;
}

class Shape
{
public:
    virtual string GetInfo() = 0;
    virtual void MoveBy(int a, int b) = 0;
};

// Implement the RectShape and TextShape descendant classes
class RectShape : public Shape
{
public:
    RectShape(int x1, int y1, int x2, int y2) : x1(x1), y1(y1), x2(x2), y2(y2) { }
    string GetInfo() override { return "R(" + to_string(x1) + "," + to_string(y1) + ")(" + to_string(x2) + "," + to_string(y2) + ")"; }
    void MoveBy(int a, int b) override { x1 += a; x2 += a; y1 += b; y2 += b; }
private:
    int x1, y1, x2, y2;
};

class TextShape : public Shape
{
public:
    TextShape(TextView& tview, int x1, int y1, int x2, int y2);
    string GetInfo() override;
    void MoveBy(int a, int b) override;
private:
    TextView& tview;
};

TextShape::TextShape(TextView& tview, int x1, int y1, int x2, int y2) : tview(tview)
{
    this->tview.SetOrigin(x1, y1);
    this->tview.SetSize(x2 - x1, y2 - y1);
}
string TextShape::GetInfo()
{
    int x0, y0, width, height;
    tview.GetOrigin(x0, y0);
    tview.GetSize(width, height);
    string x1, y1, x2, y2;
    x1 = to_string(x0);
    y1 = to_string(y0);
    x2 = to_string(x0 + width);
    y2 = to_string(y0 + height);
    return "T(" + x1 + "," + y1 + ")(" + x2 + "," + y2 + ")";
}
void TextShape::MoveBy(int a, int b)
{
    int x0, y0;
    tview.GetOrigin(x0, y0);
    tview.SetOrigin(x0 + a, y0 + b);
}

void Solve()
{
    Task("OOP2Struc3");
    int n, a, b;
    pt >> n;
    vector<Shape*> shape(n);
    vector<TextView> text(n, TextView());
    for (int i = 0; i < n; i++) {
        char c;
        pt >> c;
        int x_1, y_1, x_2, y_2;
        pt >> x_1 >> y_1 >> x_2 >> y_2;
        if (c == 'R')
            shape[i] = new RectShape(x_1, y_1, x_2, y_2);
        else
            shape[i] = new TextShape(text[i], x_1, y_1, x_2, y_2);
    }
    pt >> a >> b;
     for (auto i : shape) {
        i->MoveBy(a, b);
        pt << i->GetInfo();
    }
}

File: OOP2Struc3.cs
// File: "OOP2Struc3"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public class TextView
        {
            // Do not change the implementation of the class
            int x, y;
            int width = 1, height = 1;
            public void GetOrigin(out int x, out int y)
            {
                x = this.x;
                y = this.y;
            }
            public void SetOrigin(int x, int y)
            {
                this.x = x;
                this.y = y;
            }
            public void GetSize(out int width, out int height)
            {
                width = this.width;
                height = this.height;
            }
            public void SetSize(int width, int height)
            {
                this.width = width;
                this.height = height;
            }
        }
        public abstract class Shape
        {
            public abstract string GetInfo();
            public abstract void MoveBy(int a, int b);
        }

        // Implement the RectShape and TextShape descendant classes
        public class RectShape : Shape
        {
            public RectShape(int x1, int y1, int x2, int y2)
            {
                this.x1 = x1;
                this.y1 = y1;
                this.x2 = x2;
                this.y2 = y2;
            }
            public override string GetInfo() { return "R(" + Convert.ToString(x1) + "," + Convert.ToString(y1) + ")(" + Convert.ToString(x2) + "," + Convert.ToString(y2) + ")"; }
            public override void MoveBy(int a, int b) { x1 += a; x2 += a; y1 += b; y2 += b; }
            private int x1, y1, x2, y2;
        }
        public class TextShape : Shape
        {
            public TextShape(TextView tview, int x1, int y1, int x2, int y2)
            {
                this.tview = tview;
                this.tview.SetOrigin(x1, y1);
                this.tview.SetSize(x2 - x1, y2 - y1);
            }
            public override string GetInfo()
            {
                tview.GetOrigin(out int x0, out int y0);
                tview.GetSize(out int width, out int height);
                return "T(" + Convert.ToString(x0) + "," + Convert.ToString(y0) + ")(" + Convert.ToString(x0 + width) + "," + Convert.ToString(y0 + height) + ")";
            }
            public override void MoveBy(int a, int b)
            {
                tview.GetOrigin(out int x0, out int y0);
                tview.SetOrigin(x0 + a, y0 + b);
            }
            private TextView tview;
        }

        public static void Solve()
        {
            Task("OOP2Struc3");
            int n = GetInt(), a, b;
            var shape = new Shape[n];
            var text = new TextView[n];
            for (int i = 0; i < n; i++)
            {
                char c = GetChar();
                int x_1, y_1, x_2, y_2;
                (x_1, y_1, x_2, y_2) = GetInt4();
                text[i] = new TextView();
                if (c == 'R')
                    shape[i] = new RectShape(x_1, y_1, x_2, y_2);
                else
                    shape[i] = new TextShape(text[i], x_1, y_1, x_2, y_2);
            }
            (a, b) = GetInt2();
            foreach (var i in shape)
            {
                i.MoveBy(a, b);
                Put(i.GetInfo());
            }
        }
    }
}
